### PR TITLE
Fix/perm full parametrisation

### PR DESCRIPTION
**Pull request by kalmarek**
```julia
using AbstractAlgebra
import AbstractAlgebra.Generic: CycleDec, cycledec
using BenchmarkTools
function perm_sample(N::T=1000; samplesize=1000, seed=1234) where T<:Integer
    srand(seed)
    G = PermutationGroup(N)
    return [rand(G) for i in 1:samplesize]
end

ps = perm_sample(N)
function test_pow(ps)
    k = 0
    for p in ps 
        pp = p^6
        k+= (pp)[1]
        cycledec(pp.d)
    end
    k
end
@benchmark test_pow(ps)
```
`N = 1000`,  old:
```
  memory estimate:  25.15 MiB
  allocs estimate:  35408
  --------------
  minimum time:     17.508 ms (7.32% GC)
  median time:      17.784 ms (7.39% GC)
  mean time:        18.480 ms (8.37% GC)
  maximum time:     31.734 ms (17.22% GC)
  --------------
  samples:          271
  evals/sample:     1
```
`N = 1000`, new:
```
  memory estimate:  16.76 MiB
  allocs estimate:  25621
  --------------
  minimum time:     9.265 ms (0.00% GC)
  median time:      11.123 ms (12.76% GC)
  mean time:        10.914 ms (10.17% GC)
  maximum time:     13.668 ms (11.86% GC)
  --------------
  samples:          458
  evals/sample:     1
```
for `N=100` the ratio of medians is `3 (ms) : 2 (ms)`; for `N=10000` it is `177:106`

this pull builds on top of the previous;

### CODE REVIEW

**Comment by kalmarek** [comment]
this one is real though ;-)

**Comment by kalmarek** [comment]
* Please have a look at `_numpart`; Any design idea to make it "more" type stable (the hot code is already stable, right?)
* I used promotion rules to make multiplication of `perm{I}` and `perm{J}` possible for all `I`, `J`. We may decide to warn the user about the promotion happening, or ask for an explicit conversion. What do You think about those options?
* I chose to compare permutations only based on the object they represent in the abstract permutation group (regardless of their type and `parent`). This makes sense mathematically, but rises the question: why do we even keep the `parent` field when `parent` is readily reconstructable from the information within a `perm`?
* (continuing the above) Since parent objects are very cheap to store, why don't we store every constructed `PermGroup`? Do you have examples of computations requiring constructions of millions of different `PermGroup`s (when caching-not caching would start to matter, maybe?) 
* There are a few corner cases when handling `UInt`s and `Ints`: e.g. `factorial(::T)` returns `Int64` for all T<:Integer, but `UInt64` (which returns an `UInt64`), and of course `BigInt`. This will mostly pronounce itself when computing `orders` or dimensions of representations. Any idea on a clean resolution? This is type-stable from the point of view of the compier, but not from the point of view of the user. should we always return `Int` or `BigInt`?
* comment on anything that You find suspicious ;-)

File: src/generic/YoungTabs.jl [hunk 1]
```diff
@@ -78,10 +78,47 @@ function Base.start(parts::AllParts{T}) where T<:Integer
...
 
+const _numPartsTable = Dict{Int, Int}(0 => 1, 1 => 1, 2 => 2)
+const _numPartsTableBig = Dict{Int, BigInt}()
+
+doc"""
+    _numpart(n::Int)
+> Returns the number of all distinct integer partitions of `n`. The function
+> uses Euler pentagonal number theorem for recursive formula. For more details
+> see OEIS sequence [A000041](http://oeis.org/A000041). Note that
+> `noPartitions(0) = 1` by convention.
```
Comment (thofma) [inline src/generic/YoungTabs.jl]: `_numpart(0) = 1`?

Comment (kalmarek) [reply src/generic/YoungTabs.jl]: Yeah, the leftovers spott the function, I'm fixing it and adding tests as well

**Comment by kalmarek** [comment]
Ok, this branch has become a mess, sorry; I decided that parametrising Partitions/YoungTableau/SkewDiags doesn't make much sense right now (I don't need anything but Int), and it is hard to imagine  applications where it does. So I reverted those changes;

The only meaningful difference would be to allow e.g. `dim(::YoungTableau)` to return `Int` or `BigInt`, based on the parameter type of `YoungTableau` but I guess this is not worth the hassle, as You can do `character(part::Partition, PermGroup{BigInt}(100)())` to obtain the correct answer.

In the end I still need to rewrite the history of the branch, so I think we can abandon this pull....

In a separate branch I will also de-parentize perms (I hope before the weekend).

**Comment by kalmarek** [comment]
The overflow will happen even for `order(::PermGroup{Int})`; But if the user is aware that it may happen, he/she can choose to use `PermGroup{BigInt}`, which will always give the correct answer, at the cost of computing with `BigInt`s. Moreover the solution
```julia
order(G::PermGroup{T}) where T = Int(factorial(G.n))
order(G::PermGroup{BigInt}) = factorial(G.n)
```
is type stable and as fast as it can get. A similar solution
```julia
dim(Y::YoungTableau{T}) where T = Int(...)
dim(Y::YoungTableau{BigInt}) = .... # returns BigInt
```
just leaves to the user to check that there is no overflow happening.

Or should we protect the user even at the cost of doing everything with `BigInt`s?

**Comment by thofma** [comment]
I don't know if I would call it protection. In practice a lot of users will use `PermGroup{Int}` and `perm{Int}`, because arithmetic etc. is so much faster.

But then all of a sudden `order(PermGroup(40))` or `order(p::perm{Int})` fail. Then just for computing the order or the dimension one has to temporarily construct everything again using `BigInt`.

I think from a users point of view, these operations should never fail due to a conversion error.

**Comment by wbhart** [comment]
Sorry to interrupt the thread. I just wanted to mention that we've
apparently decided on using the Gap convention for composition, i.e. p*q
means to first apply p, then apply q.

The reasons for this are very involved and took a very long email
discussion and discussion in real life amongst the people doing commutative
and noncommutative algebra, where the same consideration arises. The final
argument boils down to requiring the order of representation matrices to
match the order of the corresponding maps in noncommutative algebra, where
left R-modules are certainly the usual thing.

I realise this is the opposite of what we previously said, but it has been
a long road to get to this decision. I suspect the system will feel more
consistent if we apply this convention throughout the whole system,
including the Groups.

An exception to this will be permutation matrices used for permutations of
the rows of a matrix. This cannot follow the same rule and will therefore
will differ either in the order of the products or in the notation.

The Gap people in our project are also keen to stick with the Gap
convention on this one. I believe it is also the Magma convention. So there
is a precedent for it.

**Comment by kalmarek** [comment]
@wbhart I think this will be for the next pull, I have my hands full fixing this one...

**Comment by kalmarek** [comment]
@thofma Yes, I understand this point of view, but one `BigInt` spilling into the computation will slow it significantly, without signalling the reason. 
This is the precise reason I don't think this is reasonable to always return `BigInt` even if `Int` would suffice.

If `order(p)` overflows for a `p::perm{Int}` it is enough to call `order(convert(perm{BigInt}, p))` to get the correct answer, or `order(PermGroup(big(length(p.d)))`. Not such a large price to pay for stability, I think.

On the second thought, if we insist on the correct-or-error behaviour we could `using Checked` and call e.g. `reduce(checked_mul, 1, 1:40)` for order(::PermGroup{<:Int}). 

Btw. none of the bechaviours I am particularly attached to (my use cases are just to small), just exploring different solutions/opinions.

**Comment by kalmarek** [comment]
> I think having a second version `order(p::perm{T}, Type{S})`, where you can specify the return type via `S`, is the better approach. The default behavior should be "it is working", and the behavior "I want it to be fast but 'broken' for wrong input" should be optional.

This is a fair solution, I'm happy to adopt it. Is `func(arg, outputType::Type{S})` a common idiom in AbstractAlgebra? So far I came up with
```julia
order(G::PermGroup, ::Type{T}=BigInt) where T = factorial(T(G.n))

doc"""
    order(a::perm, T::Type)
> Returns the order of permutation `a` as `T`, or its widening.
"""
order(a::perm, ::Type{T}=BigInt) where T = reduce(lcm, one(T), diff(cycles(a).cptrs))
```
got to think abount characters though...

**Comment by wbhart** [comment]
I think it is usual to have the type first. See for example convert, which is a standard Julia function, which we also overload, e.g. for arbs.

**Comment by kalmarek** [comment]
any idea why `reduce(lcm, UInt16(1), [4])` is an `Int`, but `reduce(lcm, UInt32(1), [4])` is an `UInt64`, even though both `lcm(UInt16(1), 4)` and `lcm(UInt32(1), 4)` return an Int?

**Comment by kalmarek** [comment]
I think this is ready to merge, if unless someone has objections. 

I had to workaround the "weird promotion and widening rules", see https://github.com/Nemocas/AbstractAlgebra.jl/pull/80/commits/c154657fca076a14f785535347f41616229a3857

**Pull request merged**